packages/react/src/cache/ReactCache.js

```diff
@@ -13,15 +13,27 @@ import invariant from 'shared/invariant';
 
 type Cache = {|
   resources: Map<any, any>,
+  disposeCallbacks: Array<() => void>,
+  dispose: () => void,
 |};
 
 // TODO: should there be a default cache?
 const CacheContext: ReactContext<null | Cache> = createContext(null);
 
 function CacheImpl() {
   this.resources = new Map();
-  // TODO: cancellation token.
+  this.disposeCallbacks = [];
 }
+CacheImpl.prototype = {
+  constructor: CacheImpl,
+  dispose() {
```

Comment:
If we do have a helper, we should make it a tree-shakable method because it's possible you get stuck with a cache impl as part of your dependencies even if you don't actually use it.  So It's nicer if this is just a protocol. It's likely that you only have one usage of the cache and in that case it might be better that it's inline.

We allow arbitrary access to the Map so why not this?.